Added Solution

alinschuller opened the pull request.

alinschueller committed:
Added Solution

alexandru-calinoiu reviewed (changes requested):
Please clean up the repo and only commit meaningful files, use a .gitignore generator, for example we don't need a `bin/` folder 

Also it will be helpfull to update the readme with minimal instructions on how to run this.

Commit:
Updated README.MD file to contain minimal instruction

Commit:
added .gitignore file

Commit:
Removed unused files specified in gitignore

pirvudoru commented on INeedCoffee.Tests/UserTests.cs:
```diff
@@ -0,0 +1,25 @@
+
+            string expected = "45.7838552 24.1515281";
+
+            Assert.Equal(expected, actual);
```
Why did you choose to concatenate values and assert on that vs checking individiual properties?

alinschuller replied:
When I built the test case, I based it on the from the UserCoordinatesCapture file, but definitely should’ve checked individual properties instead of being based on that method as I’ve done in UserCoordinatesValidator

pirvudoru commented on INeedCoffee/CoffeeShop/CoffeeShop.cs:
```diff
@@ -0,0 +1,10 @@
+    {
+        public string Name { get; set; }
+        public double Latitude { get; set; }
+        public double Longitude { get; set; }
```
Seems that Latitude and Longitude always go together (ex: CoffeeShop and UserLocation). What can we do about them taking OOP into consideration?

alinschuller replied:
We can create an immutable struct for Coordinates

pirvudoru commented on INeedCoffee/CoffeeShop/CoffeeShop.cs:
```diff
@@ -0,0 +1,10 @@
+        public string Name { get; set; }
+        public double Latitude { get; set; }
+        public double Longitude { get; set; }
+        public double DistanceToUser { get; set; }
```
Can a CoffeeShop exist without a distance? Be careful when mutating objects because it's going to become a mess if we loose track of where the mutations are made, which in turn will create unwanted, hard to debug behavior.

alinschuller replied:
Yes, it definitely can exist without a distance and it makes sense it will cause issues if we want to scale up our application. I will make sure to document myself properly to avoid situations like this.

pirvudoru commented on INeedCoffee/Utilities/HaversineDistanceCalculator.cs:
```diff
@@ -0,0 +1,20 @@
+    {
+        public static double Calculate(double firstLatitude, double firstLongitude, double secondLatitude, double secondLongitude)
+        {
+            var earthRadiusInKm = 6372.8;
```
Problem statement says:

```
Assume all coordinates lie on a plane.
```

alinschuller replied:
In this case, we should return the following formula: 
- Math.Sqrt(Math.Pow((firstLatitude - secondLatitude), 2) + Math.Pow((firstLongitude - secondLongitude), 2);

alexandru-calinoiu replied:
You are right,

But what is the problem with your solution, given that you implemented haversine but problem stated something simpler? (it goes by a specific name and it's usually considered a bad thing)

alinschuller replied:
It's definitely the fact that I used static

pirvudoru commented on INeedCoffee/Validation/CSVFilePathValidator.cs:
```diff
@@ -0,0 +1,22 @@
+{
+    public static class CSVFilePathValidator
+    {
+        public static bool Validate(string[] csvFileName)
```
Can you name Pros / Cons of using bool as a return type? What about passing a bool as a parameter to a method?

alinschuller replied:
**Can you name Pros / Cons of using bool as a return type?**
PROS:
- Makes a function easier to read and evaluate
- Useful when a statement has only two possible outcomes 

CONS: 
- We cannot return any value other than True or False 
- Returning True or False does not give much information 

**What about passing a bool as a parameter to a method?**
Passing a bool as a parameter to a method can represent a code smell. One method can do only one thing and if it contains a bool parameter, most probably it means it does two things. 
Also, if we want to add in the future new conditions, this means that we will need to add new parameters which will make it very hard to maintain

pirvudoru commented on INeedCoffee/Validation/TerminalArgumentsValidator.cs:
```diff
@@ -0,0 +1,33 @@
+            }
+
+            return true;
+            }
```
File does not look formatted correctly. 

https://coryrylan.com/blog/why-enforcing-code-style-is-important

alinschuller replied:
I formatted most of the files and most probably skipped this one.

pirvudoru commented on INeedCoffee/Validation/CSVFilePathValidator.cs:
```diff
@@ -0,0 +1,22 @@
+
+namespace INeedCoffee
+{
+    public static class CSVFilePathValidator
```
What is the purpose of having multiple concrete Validators without the ability to interchange them? Would think this is not the pattern we want to use for our implementation. Probably you were looking for something like http://www.stavroskasidis.com/blog/2017/tips-and-tricks-1-guard-clauses/

alinschuller replied:
Reading the tips and tricks from the link that you sent me and also reading the post from this link that I just found [(),](https://www.siepman.nl/blog/post/2014/01/05/SOLID-validation-class-with-easy-logical-implication.aspx) made me understand why my approach is not a good one and showed me a better one

alinschuller replied:
Reading the tips and tricks from the link that you sent me and also reading the post from this link that I just found ([https://www.siepman.nl/blog/post/2014/01/05/SOLID-validation-class-with-easy-logical-implication.aspx](url)), made me understand why my approach is not a good one and showed me a better one

alinschuller replied:
Reading the tips and tricks from the link that you sent me and also reading the post from this link that I just found ([https://www.siepman.nl/blog/post/2014/01/05/SOLID-validation-class-with-easy-logical-implication.aspx](https://www.siepman.nl/blog/post/2014/01/05/SOLID-validation-class-with-easy-logical-implication.aspx)), made me understand why my approach is not a good one and showed me a better one

pirvudoru commented on INeedCoffee/CoffeeShop/ClosestCoffeeShopsFinder.cs:
```diff
@@ -0,0 +1,21 @@
+{
+    public static class ClosestCoffeeShopsFinder
+    {
+        public static void Find(List<CoffeeShop> coffeeShopsList, int numberOfClosestCoffeeShopsToShow, int numberOfDecimalsToRoundDistanceTo)
```
Does not look too OOP. We are overusing static methods
https://stackoverflow.com/questions/2671496/java-when-to-use-static-methods

alinschuller replied:
I had a moment in which I used the Visual Studio’s fix all utility without focusing on each method to define it properly

pirvudoru reviewed (changes requested):
Please answer the questions in my review :) Thanks!

alexandru-calinoiu commented:
Clone the project, run tests, and I get 3 failures

![Screenshot from 2019-04-11 21-51-07](https://user-images.githubusercontent.com/679198/55983504-f6b51d00-5ca3-11e9-8591-4483a8d62223.png)

Most likely there is a setup step that I am missing.

alexandru-calinoiu commented on INeedCoffee/INeedCoffee.cs:
```diff
@@ -0,0 +1,29 @@
+
+namespace INeedCoffee
+{
+    class INeedCoffee
```
Is a funny class name, to bad that .net has this convention to prefix interfaces with `I`, so my instinct first time was what the ....

alinschuller replied:
I realized that as well but somehow I liked the name so I just said "nah, I'll stick with it"

alinschuller replied:
Yes, indeed, I just realized at a point that it looks like an Interface but I did not want to use MeWantCoffee, so I just said "nah, I'll stick with this even if it's weird"

alexandru-calinoiu reviewed (commented):
The solution get's the job done, is better that the initial solution, but it still suffers form tight coupling between classes, it has a bunch of other [code smells](https://blog.codinghorror.com/code-smells/) and the unit tests are rather sparse.

alinschuller commented:
Hi @alexandru-calinoiu , the failures occur because the file is not in that path that it specifies. I had the file initially in the Debug folder which I added in .gitignore and removed it through a couple of git commands. Therefore, the person who is missing a step is me, I should've referenced a path in which the file always existed and not only in the current folder. If you run the tests via terminal from the folder INeedCoffee_ReadyToRun, they should work because they take the file from the current folder.

alexandru-calinoiu commented:
Thank you, will close this and followup via email.
